### PR TITLE
[FLINK-8620] Enable shipping custom files to BlobStore and accessing them through DistributedCache

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

*  This change allows distributing any custom files to TaskManagers through BlobServer. Files can be accessed through DistributedCache in TaskManagers.

## Brief change log
  - Files are uploaded to BlobServer during submission via RestClient
  - Blob keys for those files are distributed via JobConfig
  - FileCache accesses also files from BlobService based on DistributedCacheEntry

## Verifying this change

- I've added end-to-end test `test-infra/end-to-end-test/test_streaming_distributed_cache_via_blob.sh`
- Added unit test for integration FileCache <-> BlobService

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no 
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: yes 
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? not documented

### CODE REVIEW

**Comment by dawidwys** [comment]
R: @aljoscha

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 1]
```diff
@@ -314,6 +315,9 @@ public static TaskExecutor startTaskManager(
 
 		TaskManagerConfiguration taskManagerConfiguration = TaskManagerConfiguration.fromConfiguration(configuration);
 
+		final FileCache fileCache = new FileCache(taskManagerServicesConfiguration.getTmpDirPaths(),
```
Comment (ifndef-SleePy) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: It seems that the `fileCache` here is not been used at all.

**Comment by zentol** [comment]
As far as i know one of the end-goals of FLIP-6 is to have all client-cluster communication go through the REST API. This implies that the client cannot submit things directly to the blob service.

I'm wondering whether it really makes sense to change the current behavior as any change made would effectively be deprecated-on-arrival.

**Comment by ifndef-SleePy** [comment]
Hi @zentol 

In Flip-6 the client communicates with cluster via REST API, that's true. However this not include blobs. Currently in `RestClusterClient.submit()` method, the client uses blob service to upload jars of user. So this PR does not break the rule.

I think this feature is quite good. Spark, even MapReduce support uploading user-defined jars/files/archives. I think we should support it. Users would be able to migrate to Flink more fluently with this feature.

What do you think?

**Comment by ifndef-SleePy** [comment]
If some day the behavior of `RestClusterClient` changes. Maybe uploading the jars via REST API. We could just upload the user-defined files via REST API either. That's not a big problem, right?

**Comment by zentol** [comment]
I'm aware that the `RestClusterClient` currently submits jars directly to the blob service. If this PR would've been opened half a year ago I would've welcomed it with open arms.

My point still stands though, every line that is touched here would have to be rewritten once we do that, which may very well happen before 1.6.

**Comment by dawidwys** [comment]
@zentol I agree it will need to be changed in the end (though wouldn't say rewritten completely) after the submission process is completed. 

I think though this change is important from the perspective of integration with portability layer in beam. Would be glad to hear @aljoscha's opinion on that.

**Comment by ifndef-SleePy** [comment]
@zentol All right, got your point. That's a problem indeed.

File: flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java [hunk 2]
```diff
@@ -106,9 +106,9 @@
  * are created for the plan nodes, on the way back up, the nodes connect their predecessors.
  */
 public class JobGraphGenerator implements Visitor<PlanNode> {
-	
+
 	public static final String MERGE_ITERATION_AUX_TASKS_KEY = "compiler.merge-iteration-aux";
-
```
Comment (zentol) [inline flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java]: this would be a lot easier to review without the formatting changes :/

Comment (dawidwys) [reply flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java]: Sorry for that. Forgot to disable stripping whitespaces on save. Will revert those changes.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 3]
```diff
@@ -1829,6 +1830,28 @@ public void registerCachedFile(String filePath, String name) {
...
+	 * {@link org.apache.flink.api.common.functions.RichFunction#getRuntimeContext()} and provides access
+	 * {@link org.apache.flink.api.common.cache.DistributedCache} via
+	 * {@link org.apache.flink.api.common.functions.RuntimeContext#getDistributedCache()}.
+	 *
+	 * @param filePath The path of the file
+	 * @param name The name under which the file is registered.
+	 * @param executable flag indicating whether the file should be executable
+	 * @param upload flag indicating if the file should be distributed via BlobServer
+	 */
+	public void registerCachedFile(String filePath, String name, boolean executable, boolean upload) {
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: I'm wondering whether we really need a new method. We could just modify the behavior of the existing method which should be transparent to the user.

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		}
+	}
+
+	@Test
+	public void testFileDownloadedFromBlob() {
+		try {
+			JobID jobID = new JobID();
+
+			final PermanentBlobKey permanentBlobKey = new PermanentBlobKey();
+			when(blobService.getFile(jobID, permanentBlobKey)).thenAnswer(inv -> {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java]: Given that you only have to implement 2 methods (one of which can be empty) I'd argue we don't need a mock here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+					File f = temporaryFolder.newFile("cacheFile");
+					try {
+						Files.write(testFileContent, f, Charsets.UTF_8);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java]: use `FileUtils#writeFileUtf8` instead

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+				InstantiationUtil.serializeObject(permanentBlobKey));
+			Future<Path> copyResult = fileCache.createTmpFile(fileName, entry, jobID);
+
+			final Path dstPath = copyResult.get();
+			final String actualContent =
+				StringUtils.join(Files.readLines(new File(dstPath.toUri()), StandardCharsets.UTF_8), "\n");
+			assertTrue(dstPath.getFileSystem().exists(dstPath));
+			assertEquals(testFileContent, actualContent);
+
+		} catch (Exception e) {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java]: just let the exception fail the test, i.e. no catching.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java (hunk 3 above)
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: What behaviour do you suggest sending all local files through Blob and serving files from DFS as before?

Comment (zentol) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: send everything through the blob service

**Comment by dawidwys** [comment]
@zentol I've tried changing the behaviour of `registerCachedFile` to always distribute files via BlobServer, but I've stumbled across problem with python API. Right now whole directories can be distributed via DFS. Unfortunately we can upload only single files under a single key to BlobServer. I wonder what should be better solution for it:

1. Extend BlobClient to enable uploading whole directories under a single key
2. Upload python directories as plain files (with apropriately adjusted keys e.g. path/to/filename) and later on restore structure on access

What do you think?

**Comment by aljoscha** [comment]
So the current state of this PR doesn't work with the Python API because it uploads complete directories? Is that really used/needed by the Python API?

I would like to have this feature in 1.5, if possible. This PR also adds an end-to-end test, which should make it very robust against future changes.

**Comment by dawidwys** [comment]
@aljoscha Yes that's right. The latest commit does not work because Python API expects complete directories to be uploaded. I think it is quite important cause this way the uploaded Python code can be modularized and python libs can be used. Without the directories only a single file scripts could be executed. If I understood it correctly.

Version from the commit a774fb664deabb6fdb437546038fc9477d291eb1 works, but Python API still uses DFS underneath DistributedCache.

**Comment by zentol** [comment]
Not supporting directories would be a regression regardless of whether the Python API needs it or not.

We either have to add a new method with different behavior, or zip directories underneath before uploading them.

**Comment by dawidwys** [comment]
I am not suggesting dropping uploading directories, but rather gradually switch to blob server. Leave the possibility to distribute files via DFS (and leave Python using it) and only for 1.6.0 drop it completely.

But of course I am open to suggestions.

**Comment by aljoscha** [comment]
Ok, to unblock this, I would suggest to add the new functionality under a different name, support only adding a single file (and also verify that the path passed in is a single file). @dawidwys would that be straightforward to do?

**Comment by aljoscha** [comment]
I think that's a very good solution in the end that even simplifies things. What do you think, @zentol ?

**Comment by aljoscha** [comment]
@zentol What do you think we still need now? I think this makes things easier for users. And we also need this for work on the Beam Flink Runner.

File: flink-libraries/flink-python/src/test/java/org/apache/flink/python/api/PythonPlanBinderTest.java [hunk 7]
```diff
@@ -30,6 +33,7 @@
 /**
  * Tests for the PythonPlanBinder.
  */
+@Category(Flip6.class)
```
Comment (zentol) [inline flink-libraries/flink-python/src/test/java/org/apache/flink/python/api/PythonPlanBinderTest.java]: why does this not run anymore against legacy clusters?

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java [hunk 8]
```diff
@@ -527,59 +403,83 @@ else if (response == RETURN_ERROR) {
...
-			throw new IOException("Could not transfer error message", e);
+	}
+
+	private static void compressDirectoryToZipfile(FileSystem fs, FileStatus rootDir, FileStatus sourceDir, ZipOutputStream out) throws IOException {
+		for (FileStatus file : fs.listStatus(sourceDir.getPath())) {
+			LOG.info("Zipping file: {}", file);
+			if (file.isDir()) {
+				compressDirectoryToZipfile(fs, rootDir, file, out);
+			} else {
+				String entryName = file.getPath().getPath().replace(rootDir.getPath().getPath(), "");
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java]: coudln't you used `Path#getName()`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java [hunk 9]
```diff
@@ -267,101 +208,60 @@ private static Thread createShutdownHook(final FileCache cache, final Logger log
...
-	private static class DeleteProcess implements Runnable {
+	private static class CopyFromBlobProcess implements Callable<Path> {
 
-		private final Object lock;
-		private final Map<JobID, Map<String, Tuple4<Integer, File, Path, Future<Path>>>> entries;
-
-		private final String name;
+		private final PermanentBlobKey blobKey;
+		private final Path target;
+		private final boolean directory;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: -> `isDirectory`

File: flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java [hunk 10]
```diff
@@ -267,101 +208,60 @@ private static Thread createShutdownHook(final FileCache cache, final Logger log
...
-										parent.delete();
-									}
-								}
+		public Path call() throws IOException {
+			final File file = blobService.getFile(jobID, blobKey);
+
+			if (directory) {
+				try (ZipInputStream zis = new ZipInputStream(new FileInputStream(file))) {
+					ZipEntry entry;
+					while ((entry = zis.getNextEntry()) != null) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: the entry should be closed by calling `zis.closeEntry()`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java [hunk 11]
```diff
@@ -267,101 +208,60 @@ private static Thread createShutdownHook(final FileCache cache, final Logger log
...
-
-		private final String name;
+		private final PermanentBlobKey blobKey;
+		private final Path target;
+		private final boolean directory;
+		private final boolean executable;
 		private final JobID jobID;
+		private final PermanentBlobService blobService;
 
-		public DeleteProcess(Object lock, Map<JobID, Map<String, Tuple4<Integer, File, Path, Future<Path>>>> entries,
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: who's responsible for cleaning up the created local files?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java [hunk 12]
```diff
@@ -453,15 +460,51 @@ private void testGetFailsDuringStreaming(@Nullable final JobID jobId, BlobKey.Bl
...
+				inetAddress, getBlobClientConfig())) {
+
+			key = client.uploadFile(jobId, new Path(newFolder.getPath()));
+		}
+
+		final File file = getBlobServer().getFile(jobId, (PermanentBlobKey) key);
+
+		try (ZipInputStream zis = new ZipInputStream(new FileInputStream(file))) {
+			ZipEntry entry;
+			while ((entry = zis.getNextEntry()) != null) {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java]: entry should be closed

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java [hunk 13]
```diff
@@ -453,15 +460,51 @@ private void testGetFailsDuringStreaming(@Nullable final JobID jobId, BlobKey.Bl
...
+				String fileName = entry.getName().replaceAll("/", "");
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java]: `Path#getName()`?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java [hunk 14]
```diff
@@ -453,15 +460,51 @@ private void testGetFailsDuringStreaming(@Nullable final JobID jobId, BlobKey.Bl
...
 	@Test
 	public void testUploadJarFilesHelper() throws Exception {
 		uploadJarFile(getBlobServer(), getBlobClientConfig());
 	}
 
+	@Test
+	public void testDirectoryUploading() throws IOException {
+		final File newFolder = temporaryFolder.newFolder();
+		final File file1 = File.createTempFile("pre", "suff", newFolder);
+		FileUtils.writeStringToFile(file1, "Test content");
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java]: we have somewhere in a `FileUtils` a writeUtf8 method

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java [hunk 15]
```diff
@@ -19,30 +19,38 @@
...
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
 
 /**
- * Test delete process of {@link FileCache}. The local cache file should not be deleted why another task comes in 5 seconds.
+ * Tests that {@link FileCache} can read files from {@link BlobServer}.
  */
-public class FileCacheDeleteValidationTest {
+@RunWith(MockitoJUnitRunner.class)
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java]: Why is this needed now? I don't see any mocking being added.

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java [hunk 16]
```diff
@@ -57,89 +65,59 @@
...
+		// copy / create the file
+		final DistributedCache.DistributedCacheEntry entry = new DistributedCache.DistributedCacheEntry(
+			fileName,
+			false,
+			InstantiationUtil.serializeObject(permanentBlobKey));
+		Future<Path> copyResult = fileCache.createTmpFile(fileName, entry, jobID);
+
+		final Path dstPath = copyResult.get();
+		final String actualContent =
+			StringUtils.join(Files.readLines(new File(dstPath.toUri()), StandardCharsets.UTF_8), "\n");
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java]: use `Files.readLines(new File(dstPath.toUri()).collect(Collections.joining("\n")` instead?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java (hunk 12 above)
Comment (dawidwys) [reply flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java]: It does not need to. `getNextEntry` does it, at the very beginning.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java (hunk 8 above)
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java]: Unfortunately I could not it removes only the part to the root of the directory. If we have nested directories we need more than just `getName`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java (hunk 13 above)
Comment (dawidwys) [reply flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java]: It is not sufficient. In this testcase `entry.getName` returns `/pre,,,suff`. Note the `/` in the beginning.

Though I've changed `replaceAll` to `replaceFirst`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java (hunk 10 above)
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: It does not need to. getNextEntry does it, at the very beginning.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: what about the last entry?

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: however one could argue that this is an implementation detail.

File: flink-tests/src/test/java/org/apache/flink/test/iterative/aggregators/AggregatorsITCase.java [hunk 17]
```diff
@@ -91,6 +93,7 @@ public void after() throws Exception{
 	}
 
 	@Test
+	@Category(Flip6.class)
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/test/iterative/aggregators/AggregatorsITCase.java]: revert?

File: flink-tests/src/test/java/org/apache/flink/test/distributedcache/DistributedCacheTest.java [hunk 18]
```diff
@@ -40,6 +42,7 @@
 /**
  * Test the distributed cache.
  */
+@Category(Flip6.class)
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/test/distributedcache/DistributedCacheTest.java]: revert?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 19]
```diff
@@ -1831,4 +1831,5 @@ public void registerCachedFile(String filePath, String name) {
 	public void registerCachedFile(String filePath, String name, boolean executable) {
 		this.cacheFile.add(new Tuple2<>(name, new DistributedCache.DistributedCacheEntry(filePath, executable)));
 	}
+
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java]: revert

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java (hunk 12 above)
Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java]: that's an implementation detail though

File: flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java [hunk 20]
```diff
@@ -143,30 +160,23 @@ public void shutdown() {
...
-			Map<String, Tuple4<Integer, File, Path, Future<Path>>> jobEntries = entries.get(jobID);
-			if (jobEntries == null) {
-				jobEntries = new HashMap<String, Tuple4<Integer, File, Path, Future<Path>>>();
-				entries.put(jobID, jobEntries);
-			}
+			Map<String, Future<Path>> jobEntries = entries.computeIfAbsent(jobID, k -> new HashMap<>());
+			final Set<ExecutionAttemptID> refHolders = jobRefHolders.computeIfAbsent(jobID, id -> new HashSet<>());
+			refHolders.add(executionId);
 
 			// tuple is (ref-count, parent-temp-dir, cached-file-path, copy-process)
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: outdated

**Comment by dawidwys** [comment]
@zentol I've rebased this PR. Feel free to review whenever you find a while.

File: flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java [hunk 21]
```diff
@@ -262,106 +216,120 @@ private static Thread createShutdownHook(final FileCache cache, final Logger log
...
-			this.cachedPath = cachedPath;
+		CopyFromBlobProcess(DistributedCacheEntry e, JobID jobID, PermanentBlobService blobService, Path target) {
+			try {
+				this.isExecutable = e.isExecutable;
+				this.isDirectory = e.isZipped;
+				this.jobID = jobID;
+				this.blobService = blobService;
+				this.blobKey = InstantiationUtil.deserializeObject(e.blobKey, Thread.currentThread().getContextClassLoader());
+				this.target = target;
+			} catch (Exception ex) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: the constructor signature should be modified instead

File: flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java [hunk 22]
```diff
@@ -262,106 +216,120 @@ private static Thread createShutdownHook(final FileCache cache, final Logger log
...
+
+		synchronized (lock) {
+			Set<ExecutionAttemptID> jobRefCounter = jobRefHolders.get(jobId);
+
+			if (jobRefCounter == null || jobRefCounter.isEmpty()) {
+				LOG.warn("improper use of releaseJob() without a matching number of createTmpFiles() calls for jobId " + jobId);
+				return;
+			}
+
+			jobRefCounter.remove(executionId);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: shouldn't we also remove the entries? Otherwise the `entries` map will continue to grow until either the TM shuts down or the or crashes with an OOM error.

File: flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java [hunk 23]
```diff
@@ -262,106 +216,120 @@ private static Thread createShutdownHook(final FileCache cache, final Logger log
...
 					}
 				}
-			}
-			catch (IOException e) {
+			} catch (IOException e) {
 				LOG.error("Could not delete file from local file cache.", e);
 			}
 		}
 	}
+
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: revert

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java [hunk 24]
```diff
@@ -631,14 +631,14 @@ else if (current == ExecutionState.CANCELING) {
...
 				{
 					LOG.info("Obtaining local cache file for '{}'.", entry.getKey());
-					Future<Path> cp = fileCache.createTmpFile(entry.getKey(), entry.getValue(), jobId);
+					Future<Path> cp = fileCache.createTmpFile(entry.getKey(), entry.getValue(), jobId, executionId);
 					distributedCacheEntries.put(entry.getKey(), cp);
 				}
 			}
 			catch (Exception e) {
 				throw new Exception(
-					String.format("Exception while adding files to distributed cache of task %s (%s).", taskNameWithSubtask, executionId),
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java]: revert

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+
+		fileCache.releaseJob(jobID, attemptID2);
+		// still should be available, file will be deleted after 5 seconds
+		assertTrue(fileStatus.isDir());
+		assertTrue(fs.exists(cacheFile));
+
+		// after a while, the file should disappear
+		long deadline = System.currentTimeMillis() + 10000;
+		do {
+			Thread.sleep(5500);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java]: well, I'd rather not blow 5-10 seconds on a single test.

I would suggest shutting down the FileCache instead. If the delete process was triggered it should still delete the file.

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java [hunk 26]
```diff
@@ -33,16 +38,16 @@
...
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
 
 /**
- * Test delete process of {@link FileCache}. The local cache file should not be deleted why another task comes in 5 seconds.
+ * Tests that {@link FileCache} can read files from {@link BlobServer}.
  */
-public class FileCacheDeleteValidationTest {
+public class FileCacheReadsFromBlobTest {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java]: am i missing something, or isn't this now a duplicate of `FileCacheDirectoryTest#testDirectoryDownloadedFromBlob`

Comment (dawidwys) [reply flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java]: It is different in the way that this test uploads a single file. Therefore it follows a bit different path. The file is not zipped and is directly shipped from BlobCache. In case of directories there is unzipped version which lifecycle is managed by `FileCache`.

If you think though the test `FileCacheDirectoryTest#testDirectoryDownloadedFromBlob` is enough, I will remove it.

Comment (dawidwys) [reply flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java]: it does, it is `FileCache` responsibility to unzip.

Comment (zentol) [reply flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheReadsFromBlobTest.java]: ah yes, i was misreading the test...
´

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java (hunk 25 above)
Comment (dawidwys) [reply flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java]: I can not just use `fileCache.shutdown` because the shutdown performs additional clearing of temporary directories, which is not what I want to test. I've changed the implementation. I am not sure thought about making the `ExecutorService` `VisibleForTesting`. Another alternative would be to revert to a similar loop, but this time with a shorter period.

File: flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java (hunk 22 above)
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/filecache/FileCache.java]: You are right! I added removing it in the DeleteProcess.

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java [hunk 27]
```diff
@@ -164,16 +165,13 @@ public void testDirectoryCleanUp() throws Exception {
 		assertTrue(fs.exists(cacheFile));
 
 		fileCache.releaseJob(jobID, attemptID2);
-		// still should be available, file will be deleted after 5 seconds
+		// still should be available, file will be deleted after 200 milliseconds
 		assertTrue(fileStatus.isDir());
 		assertTrue(fs.exists(cacheFile));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java]: what you _could_ also do is pass in a `ScheduledExecutorService`. You could then intercept the delete process, verify the interval arguments and fire the process at your leisure.

I'm worried that in it's current form the test is either not stable or takes longer than it should.

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java [hunk 28]
```diff
@@ -0,0 +1,209 @@
...
+
+		final Path cacheFile = new Path(dstPath, "cacheFile");
+		assertTrue(fs.exists(cacheFile));
+		final String actualContent = FileUtils.readFileUtf8(new File(cacheFile.getPath()));
+		assertEquals(testFileContent, actualContent);
+	}
+
+	@Test
+	public void testDirectoryCleanUp() throws Exception {
+		fileCache.shutdown();
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java]: Could we not always pass a `DeleteCapturingDirectScheduledExecutorService` into the FileCache? We can explicitly call run in `@After`; that would allow us to share the filecache between both tests again.

File: flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java [hunk 29]
```diff
@@ -0,0 +1,209 @@
...
+		final Path cacheFile = new Path(dstPath, "cacheFile");
+		assertTrue(fileStatus.isDir());
+		assertTrue(fs.exists(cacheFile));
+
+		fileCache.releaseJob(jobID, attemptID1);
+		// still should be available
+		assertTrue(fileStatus.isDir());
+		assertTrue(fs.exists(cacheFile));
+
+		fileCache.releaseJob(jobID, attemptID2);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java]: are we at risk of race-conditions here between this call and the second `createTmpFile` call? We never wait for the returned future.

Comment (dawidwys) [reply flink-runtime/src/test/java/org/apache/flink/runtime/filecache/FileCacheDirectoriesTest.java]: We are not, cause the `DirectExecutorService` is synchronous for all non-scheduled tasks. The only scheduled task is the `DeleteProcess` though, which we capture and run manually.

**Comment by dawidwys** [comment]
I've addressed @zentol last comments and rebased to fix conflicts. Will merge after travis gives green.
